pkg/rslog/debug.go

```diff
@@ -6,106 +6,125 @@ import (
+// debugRegions relates a numerical product debug region to its string
+// equivalent. Used for display (when logging) and when processing
+// configuration values.
+//
+// Protected by debugMutex.
+var debugRegions map[ProductRegion]string = map[ProductRegion]string{}
 
-var regionCallbacks = map[ProductRegion]callbacksArr{}
-var regionsEnabled = map[ProductRegion]bool{}
-var debugMutex sync.RWMutex
+// debugEnabled records product debug regions which enable debug logging.
+//
+// Protected by debugMutex.
+var debugEnabled map[ProductRegion]bool = map[ProductRegion]bool{}
 
+// RegisterRegions registers product debug regions. Called once at program
+// startup.
 func RegisterRegions(regions map[ProductRegion]string) {
-	regionNames = regions
+	debugMutex.Lock()
+	defer debugMutex.Unlock()
+
+	debugRegions = regions
 }
 
+// RegionNames returns the names of all registered product debug regions.
 func RegionNames() []string {
+	debugMutex.RLock()
+	defer debugMutex.RUnlock()
+
 	var names []string
-	for _, name := range regionNames {
+	for _, name := range debugRegions {
 		names = append(names, name)
 	}
 	return names
 }
 
+// Regions returns the numerical product debug regions.
 func Regions() []ProductRegion {
+	debugMutex.RLock()
+	defer debugMutex.RUnlock()
+
 	var regions []ProductRegion
-	for r := range regionNames {
+	for r := range debugRegions {
 		regions = append(regions, r)
 	}
 	return regions
 }
 
-func RegionByName(text string) ProductRegion {
-	for region, name := range regionNames {
-		if name == text {
+// RegionByName returns the numerical product debug region associated with
+// name. Returns zero when there is no match.
+func RegionByName(name string) ProductRegion {
+	debugMutex.RLock()
+	defer debugMutex.RUnlock()
+
+	for region, region_name := range debugRegions {
+		if name == region_name {
 			return region
 		}
 	}
 	return 0
 }
 
+// RegionName returns the name associated with the numerical product debug
+// region. Returns an empty string when there is no match.
 func RegionName(region ProductRegion) string {
-	return regionNames[region]
+	debugMutex.RLock()
+	defer debugMutex.RUnlock()
+
+	return debugRegions[region]
 }
 
-// Register debug regions enabled.
-// This should be called as early as possible when starting an application.
+// InitDebugLogs registers a set of product debug regions as enabled.
+//
+// This should be called as early as possible when starting an application,
+// but after RegisterRegions.
 func InitDebugLogs(regions []ProductRegion) {
 	debugMutex.Lock()
 	defer debugMutex.Unlock()
 
 	// Reset enabled regions on each call.
-	regionsEnabled = make(map[ProductRegion]bool)
+	debugEnabled = make(map[ProductRegion]bool)
 	lgr := DefaultLogger()
 
 	// match debug log region to list
 	for _, region := range regions {
 		if region == 0 {
 			continue
 		}
-		regionsEnabled[region] = true
-		regionName := RegionName(region)
+		debugEnabled[region] = true
+		regionName := debugRegions[region]
```

Comment:
`InitDebugLogs()` directly accesses `debugRegions` because `RegionName()` now uses `debugMutex`; we need `InitDebugLogs()` to hold the mutex until its work is complete.